rootfs/etc/nginx/lua/util/defer.lua

```diff
@@ -0,0 +1,57 @@
+local util = require("util")
+
+local _M = {
+  timer_started = false,
+  queue = {},
+  MAX_QUEUE_SIZE = 10000
+}
+
+local function flush_queue(premature)
+  -- TODO Investigate if we should actually still flush the queue when we're
+  -- shutting down.
+  if premature then return end
+
+  local queue = _M.queue
+  _M.queue = {}
+  _M.timer_started = false
+
+  for _,v in ipairs(queue) do
+    v.func(unpack(v.args))
+  end
+end
+
+-- `to_timer_phase` will enqueue a function that will be executed in a timer
+-- context, at a later point in time. The purpose is that some APIs (such as
+-- sockets) are not available during some nginx request phases (such as the
+-- logging phase), but are available for use in timers. There are no ordering
+-- guarantees for when a function will be executed.
+function _M.to_timer_phase(func, ...)
```

Comment:
I'm not sure if I understand this function implementation. Here's what I think will happen:

t = 1:
 `to_timer_phase` gets called
 entry gets added to the queue
 `_M.timer_started` is `false` so the `flush_queue` will be run in timer phase immediately because of `0` delay
 as soon as `flush_queue` starts it sets `_M.timer_started` to `false` again and queue is reset before actually starting executing the functions from the queue (in this case only one function)

t = 2:
 `to_timer_phase` gets called
 entry gets added to the queue - again the queue will have just one item
 `_M.timer_started` is `false` so the `flush_queue` will be run in timer phase immediately because of `0` delay
 as soon as `flush_queue` starts it sets `_M.timer_started` to `false` again and queue is reset before actually starting executing the functions from the queue (in this case only one function)

and so on
so we will never buffer/batch the function execution which I expected this utility to be doing.

Reply:
cc: @csfrancis

Reply:
From the `ngx.timer.at` docs (emphasis mine):

> `0` delay can also be specified, in which case the timer will immediately expire **when the current handler yields execution**.

We effectively execute all of our middlewares within the context of a single phase handler, i.e. `rewrite`, `access_by`, etc. So what this means is that `flush_queue` will be run as soon as the handler has finished (after all the middlewares in the phase have executed).

Reply:
That's all clear @csfrancis, my question is more about buffering and use of `_M.queue`.